File: apps/shared/AgentKit/Sources/AgentKit/PushRegistration.swift
import Foundation

/// Where APNs delivers this device's address, and where it goes next.
///
/// Shared because the two apps had this twice, near-verbatim: the same
/// token-hexing, the same hold-until-signed-in field, the same `sendIfPossible`,
/// the same failure comment. Only the device label differed, so that is the one
/// thing left to the caller.
///
/// The local `Notifier` in each app covers the case where the phone is awake and
/// the app is running. This covers the case the product exists for: the app is
/// not running, the device is asleep, and an agent on a runner three time zones
/// away is stuck. Nothing local can know that — only the daemon can, and the
/// only route to a sleeping device is APNs.
@MainActor
public final class PushRegistration: ObservableObject {
    public static let shared = PushRegistration()

    /// Whether the relay has this device's address.
    ///
    /// Published rather than swallowed, because the failure is the product's
    /// central promise going quietly missing — the device is never filed, the
    /// daemon's pushes reach nobody, and the settings screen goes on saying
    /// "Notifications can reach this device."
    @Published public private(set) var registered = false
    @Published public private(set) var lastError: String?

    /// How this device names itself. Set once at launch by each app, because
    /// `UIDevice.current.name` and `Host.current().localizedName` are the only
    /// part of this that is platform-specific.
    public var label: () -> String = { "Device" }

    /// The APNs platform for this build. `apns` on Apple; the Android client
    /// will set `fcm`.
    public var platform = "apns"

    /// Whether this device wants to hear that a turn ended — the "finishes or
    /// fails" toggle, read at registration time.
    ///
    /// A closure and not a value, and set once at launch by each app the way
    /// `label` is, because the two apps store the preference under different
    /// keys and neither of them is AgentKit's to know. Defaults to true: an
    /// app that never sets it registers as wanting notifications, which is what
    /// every install has always had.
    public var notifyOnDone: () -> Bool = { true }

    /// ActivityKit's push-to-start token, when the app has one.
    ///
    /// Filed with the device rather than on a route of its own because it is
    /// another address for the same phone, and because the two arrive in either
    /// order — re-registering when it lands is one line here instead of a second
    /// piece of hold-until-signed-in bookkeeping.
    ///
    /// Nil on the Mac, permanently. There are no Live Activities there and the
    /// relay's COALESCE is what keeps that from erasing a phone's token.
    public var liveActivityStartToken: String? {
        didSet {
            guard liveActivityStartToken != oldValue else { return }
            Task { await sendIfPossible() }
        }
    }

    /// The last token APNs gave us, held until there is an account to file it
    /// under. Registration and sign-in finish in either order, and whichever is
    /// second completes the pair.
    private var token: String?

    private init() {}

    public func received(_ deviceToken: Data) {
        token = deviceToken.map { String(format: "%02x", $0) }.joined()
        Task { await sendIfPossible() }
    }

    /// Called after signing in, for the case where the token arrived first —
    /// and after the notification toggles change, for the case below.
    ///
    /// Registration runs when a push token arrives, which is at launch. A
    /// preference the relay only hears about then is one that appears to do
    /// nothing until the app happens to re-register, so the settings screen
    /// calls this on the setter. That is the same bug this whole change is
    /// about, one level up.
    public func sendIfPossible() async {
        guard let token, Account.shared.isSignedIn else { return }
        switch await Account.shared.registerDevice(
            pushToken: token,
            platform: platform,
            label: label(),
            environment: Self.environment,
            liveActivityStartToken: liveActivityStartToken,
            notifyOnDone: notifyOnDone())
        {
        case .success:
            registered = true
            lastError = nil
        case .failure(let why):
            // The lead sentence says what did not happen; the cause says why,
            // in the relay's own vocabulary. This used to be one fixed string
            // for every failure, so a phone with no network and a phone whose
            // session had expired were told the same thing.
            registered = false
            lastError = "Couldn’t tell the relay how to reach this device. \(why.message)"
        }
    }

    /// Which APNs issued this device's token: `development` or `production`.
    ///
    /// Read out of the embedded provisioning profile rather than compiled in
    /// behind `#if DEBUG`, because those two do not line up. A Release build
    /// installed from Xcode is still development-signed, and the same source
    /// through TestFlight is not — the build configuration is simply not the
    /// question being asked. The entitlement is.
    ///
    /// This matters because getting it wrong is silent. A sandbox token sent to
    /// production APNs is refused with `BadDeviceToken`; nothing crashes,
    /// nothing logs on the phone, and the relay records an attempt it believes
    /// it made correctly. The phone just never rings.
    ///
    /// Production whenever the answer cannot be found, because that is the App
    /// Store case — the one where there is no developer watching to notice.
    public static let environment: String = {
        guard let url = Bundle.main.url(forResource: "embedded", withExtension: "mobileprovision"),
            let data = try? Data(contentsOf: url)
        else { return "production" }

        // The profile is a CMS signature wrapping a plain XML plist. We are not
        // the audience for the signature — the OS already checked it, and this
        // is reading our own bundle — so this finds the payload rather than
        // pulling in Security to unwrap it properly.
        guard let start = data.range(of: Data("<plist".utf8)),
            let end = data.range(of: Data("</plist>".utf8), in: start.upperBound..<data.endIndex),
            let parsed = try? PropertyListSerialization.propertyList(
                from: data[start.lowerBound..<end.upperBound], format: nil) as? [String: Any],
            let entitlements = parsed["Entitlements"] as? [String: Any],
            let environment = entitlements["aps-environment"] as? String
        else { return "production" }
        return environment
    }()

    /// Not surfaced as an error by the caller: registration fails on the
    /// simulator and in any build without a push entitlement, neither of which
    /// is worth telling someone about, and local notifications keep working.
    public func unavailable(_ error: Error) {
        print("remote notifications unavailable: \(error.localizedDescription)")
    }
}

File: apps/android/app/src/main/java/com/farcooler/account/PushRegistration.kt
package com.farcooler.account

import android.os.Build
import com.google.firebase.messaging.FirebaseMessaging
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Where FCM delivers this device's address, and where it goes next.
 *
 * The local `Notifier` covers the case where the phone is awake and the app is
 * running. This covers the case the product exists for: the app is not running,
 * the device is asleep, and an agent on a machine three time zones away is
 * stuck. Nothing local can know that — only the daemon can, and the only route
 * to a sleeping device is FCM.
 *
 * The relay already speaks both platforms: `services/relay/src/index.ts`
 * accepts `apns` and `fcm`, and `push.ts` sends through the FCM v1 API. So
 * nothing on the server had to change for Android — this is the client half
 * that was missing.
 */
class PushRegistration(private val context: android.content.Context, private val account: Account) {

    /**
     * Whether the relay has this device's address.
     *
     * Published rather than swallowed, because the failure is the product's
     * central promise going quietly missing — the device is never filed, the
     * daemon's pushes reach nobody, and the settings screen goes on saying
     * notifications can reach this device.
     */
    private val _registered = MutableStateFlow(false)
    val registered: StateFlow<Boolean> = _registered.asStateFlow()

    private val _lastError = MutableStateFlow<String?>(null)
    val lastError: StateFlow<String?> = _lastError.asStateFlow()

    /**
     * The last token FCM gave us, held until there is an account to file it
     * under. Registration and sign-in finish in either order, and whichever is
     * second completes the pair.
     */
    private var token: String? = null

    private var scope: CoroutineScope? = null

    fun attach(scope: CoroutineScope) {
        this.scope = scope
        // A token FCM rotated while the app was closed.
        //
        // `FarCoolerMessagingService` runs in a process that may have no
        // signed-in account and no coroutine scope to file one with, so it
        // records the new address and this picks it up. Without it a rotation
        // means every later push reaches an address nobody is listening on, and
        // nothing anywhere says so.
        val preferences = context.getSharedPreferences(
            com.farcooler.notify.FarCoolerMessagingService.PREFS,
            android.content.Context.MODE_PRIVATE,
        )
        preferences
            .getString(com.farcooler.notify.FarCoolerMessagingService.KEY_PENDING_TOKEN, null)
            ?.let { pending ->
                token = pending
                preferences.edit()
                    .remove(com.farcooler.notify.FarCoolerMessagingService.KEY_PENDING_TOKEN)
                    .apply()
            }
        requestToken()
    }

    /**
     * Ask FCM for an address.
     *
     * Wrapped, because a build with no `google-services.json` has no Firebase
     * project and this throws. That is not worth telling anyone about: it is
     * the ordinary state of a checkout somebody just cloned, local
     * notifications keep working, and the settings screen says plainly that
     * pushes cannot reach this device rather than claiming they can.
     */
    private fun requestToken() {
        runCatching {
            FirebaseMessaging.getInstance().token
                .addOnSuccessListener { received(it) }
                .addOnFailureListener { unavailable(it.message) }
        }.onFailure { unavailable(it.message) }
    }

    fun received(deviceToken: String) {
        token = deviceToken
        scope?.launch { sendIfPossible() }
    }

    /** Called after signing in, for the case where the token arrived first. */
    suspend fun sendIfPossible() {
        val token = token ?: return
        if (!account.isSignedIn) return
        val ok = account.registerDevice(pushToken = token, platform = "fcm", label = label())
        _registered.value = ok
        _lastError.value =
            if (ok) null else "Could not tell the relay how to reach this device."
    }

    fun unavailable(reason: String?) {
        _registered.value = false
        _lastError.value =
            "This build has no Firebase project, so notifications cannot reach this device " +
                "while the app is closed."
    }

    /**
     * How this device names itself on the account's device list.
     *
     * The model, because Android stopped handing out user-set device names
     * without a permission and asking for one to label a row would be a poor
     * trade.
     */
    private fun label(): String = Build.MODEL.ifBlank { "Android device" }
}
